growth/too/tasks/growthdb_cgi.py

```diff
@@ -0,0 +1,125 @@
+import requests
+import json
+import numpy as np
+import astropy.units as u
+import datetime
+from celery.task import PeriodicTask
+from celery.utils.log import get_task_logger
+from celery.local import PromiseProxy
+import celery
+
+from .. import models
+
+"""
+Reminder for the relevant program names:
+decam_programidx=program_dict['DECAM GW Followup']
+em_gw_programidx=program_dict['Electromagnetic Counterparts to Gravitational Waves']
+fermi_programidx=program_dict['Afterglows of Fermi Gamma Ray Bursts']
+neutrino_programidx=program_dict['Electromagnetic Counterparts to Neutrinos']
+"""
+
+
+def get_programidx(program_name):
+    """Given a program name, it returns the programidx"""
+
+    r = requests.post('http://skipper.caltech.edu:8080/cgi-bin/growth/list_programs.cgi')
+    r.raise_for_status()
+    programs = json.loads(r.text)
+    program_dict = {p['name']:p['programidx'] for i,p in enumerate(programs)}
+
+    try:
+        return program_dict[program_name]
+    except KeyError:
+        print(f'The user {username} does not have access to the program {program_name}')
+        return None
+
+
+def get_source_autoannotations(sourceid):
+    """ Fetch a specific source's autoannotations from the GROWTH marshal and 
+    create a string with the autoannotations available. """
+    r = requests.post('http://skipper.caltech.edu:8080/cgi-bin/growth/source_summary.cgi', data={'sourceid':str(sourceid)
+}) 
+    r.raise_for_status()
+    summary=json.loads(r.text)
+    autoannotations=summary['autoannotations']
+    autoannotations_string = ""
+    for auto in autoannotations:
+        autoannotations_string = f"{autoannotations_string}; \n {auto['username']}, {auto['type']}, {auto['comment']}" 
+
+    return autoannotations_string
+
+
+def get_candidates_growth_marshal(program_name):
+    """ Query the GROWTH db for the science programs """
+
+    programidx=get_programidx(program_name)
+    if programidx==None:
+        return None
+    r = requests.post('http://skipper.caltech.edu:8080/cgi-bin/growth/list_program_sources.cgi', data={'programidx':str(programidx)})
+    r.raise_for_status()
+    sources=json.loads(r.text)
+    """ Add autoannotations """
+    for s, i in zip(sources, np.arange(len(sources))):
+        autoannotations=get_source_autoannotations(s["id"])
```

Comment:
We have to make a separate request to the GROWTH marshal to get the annotations for each event? That seems very wasteful. This could be hundreds of separate HTTP requests!

Reply:
Oh well.